README.md
# Rulers

A Flutter Widget to show Ruler with selected values

<img src="https://raw.githubusercontent.com/theshivamlko/rulers_flutter_package/master/rulers.gif" width="220" alt="Flutter Rulers" />
<br/><br/>

## Getting Started
### Include
You should ensure that you add the following dependency in your Flutter project.
```yaml
dependencies:
 rulers: "^1.0.7"
```
You should then run `flutter packages get` in terminal.

### Import
In your Dart code, to use it:
```dart
import 'package:rulers/rulers.dart';
```

### Example
In your Dart code, to use it:
```
        Container(
                  margin: const EdgeInsets.only(top: 8.0),
                  child: RulerWidget(
                    height: 100,
                    scaleBackgroundColor: Colors.yellow,
                    indicatorWidget: Image.asset(
                         'assets/marker.png',
                          color: Colors.red,
                          height: 25,
                          width: 25,
                        ),
                    largeScaleBarsInterval: 24,
                    smallScaleBarsInterval: 3,
                    lowerIndicatorLimit: 2, 
                    upperIndicatorLimit: 8,
                    barsColor: Colors.grey,
                    inRangeBarColor: Colors.green,
                    behindRangeBarColor: Colors.orangeAccent,
                    outRangeBarColor: Colors.red,
                  ),
                ),
```

## Project Created & Maintained By

### Shivam Srivastava
 
<a href="https://linkedin.com/in/theshivamlko"><img src="https://github.com/aritraroy/social-icons/blob/master/linkedin-icon.png?raw=true" width="60"></a>
<a href="https://twitter.com/navokitech"><img src="https://github.com/aritraroy/social-icons/blob/master/twitter-icon.png?raw=true" width="60"></a>
<a href="https://facebook.com/navokitech"><img src="https://github.com/aritraroy/social-icons/blob/master/facebook-icon.png?raw=true" width="60"></a>
<a href="https://instagram.com/navokitech"><img src="https://github.com/aritraroy/social-icons/blob/master/instagram-icon.png?raw=true" width="60"></a>


For help getting started with Flutter, view our online
[documentation](https://flutter.dev/docs/get-started/install).


CHANGELOG.md
## [1.0.7]

* Add Null Safety.

## [1.0.0]

* initial release.


lib/src/RulerWidget.dart
import 'package:flutter/material.dart';

class RulerWidget extends StatefulWidget {
  /// total numbers on scale
  final int limit;

  /// starting number of scale to show marker
  final int lowerLimit;

  /// ending number of scale  to show marker
  final int upperLimit;

  /// mid starting number of scale (Optional)
  final int midLimitLower;

  /// mid ending number of scale (Optional)
  final int midLimitUpper;

  /// number of small bars of scale
  final int interval;

  /// number color of bars of scale
  final Color normalBarColor;

  /// color between lowerLimit and upperLimit bars of scale (Optional)
  final Color inRangeBarColor;

  /// color between midLimitUpper and upperLimit bars of scale (Optional)
  final Color outRangeBarColor;

  /// color between midLimitLower and lowerLimit bars of scale (Optional)
  final Color behindRangeBarColor;

  /// color of scale (Optional)
  final Color scaleColor;

  /// Custom Indicator over bars of scale (Optional)
  final Widget indicatorWidget;

  /// height of scale if horizontal or width of scale if vertical (Optional)
  final double scaleSize;

  /// Scale to be horizontal or vertical ,by default horizontal
  final Axis axis;

  final ScrollController scrollController;

  RulerWidget(
      {this.limit,
      this.scaleSize,
      this.indicatorWidget,
      this.lowerLimit = 0,
      this.midLimitLower = 0,
      this.midLimitUpper = 0,
      this.upperLimit = 0,
      this.interval,
      this.normalBarColor,
      this.scaleColor,
      this.inRangeBarColor,
      this.behindRangeBarColor,
      this.outRangeBarColor,
      this.scrollController,
      this.axis = Axis.horizontal})
      : assert(limit != null),
        assert(scaleSize != null),
        assert(interval != null),
        assert(normalBarColor != null);

  @override
  _RulerWidgetState createState() => _RulerWidgetState();
}

class _RulerWidgetState extends State<RulerWidget> {
  List<Widget> scaleWidgetList;

  @override
  void initState() {
    super.initState();

    scaleWidgetList = List();
    _generateScale();
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      color: widget.scaleColor,
      height: widget.axis == Axis.horizontal ? widget.scaleSize : null,
      width: widget.axis == Axis.vertical ? widget.scaleSize : null,
      child: SingleChildScrollView(
        controller: widget.scrollController,
        scrollDirection: widget.axis,
        child: RotatedBox(
          quarterTurns: widget.axis == Axis.horizontal ? 4 : 3,
          child: Row(
            children: scaleWidgetList,
          ),
        ),
      ),
    );
  }

  void _generateScale() async {
    for (int i = 0; i < widget.limit; i++) {
      scaleWidgetList.add(RulerBarWidget(
        key: ValueKey(i),
        num: i,
        type: RulerBar.BIG_BAR,
        lowerLimit: widget.lowerLimit,
        upperLimit: widget.upperLimit,
        indicatorWidget: widget.indicatorWidget,
        midLimitLower: widget.midLimitLower,
        midLimitUpper: widget.midLimitUpper,
        midInterval: widget.interval,
        normalBarColor: widget.normalBarColor,
        inRangeBarColor: widget.inRangeBarColor,
        behindRangeBarColor: widget.behindRangeBarColor,
        outRangeBarColor: widget.outRangeBarColor,
        axis: widget.axis,
      ));
    }

    if (mounted) setState(() {});
  }
}

class RulerBarWidget extends StatelessWidget {
  final RulerBar type;

  int num;

  final int midInterval;
  final int lowerLimit;
  final int upperLimit;
  final int midLimitLower;
  final int midLimitUpper;
  final Color normalBarColor;
  final Color inRangeBarColor;
  final Color outRangeBarColor;
  final Color behindRangeBarColor;
  final Widget indicatorWidget;
  final Axis axis;

  RulerBarWidget(
      {Key key,
      this.num,
      this.type,
      this.indicatorWidget,
      this.lowerLimit,
      this.midLimitLower,
      this.midLimitUpper,
      this.upperLimit,
      this.midInterval,
      this.normalBarColor,
      this.inRangeBarColor,
      this.behindRangeBarColor,
      this.outRangeBarColor,
      this.axis})
      : super(key: key);

  final double _spacing = 5.0;
  List<Widget> _children = List();
  bool _show = false;

  @override
  Widget build(BuildContext context) {
    _getSmallBars();

    return Container(
      child: Row(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: _children,
      ),
    );
  }

  /// get color of small bar according to range
  Color _getSmallBarColor() {
    if (_getScalePosition(num) >= _getScalePosition(lowerLimit) &&
        _getScalePosition(num) < _getScalePosition(midLimitLower))
      return behindRangeBarColor;
    else if (_getScalePosition(num) >= _getScalePosition(midLimitLower) &&
        _getScalePosition(num) < _getScalePosition(midLimitUpper))
      return inRangeBarColor;
    else if (_getScalePosition(num) > _getScalePosition(midLimitUpper) &&
        _getScalePosition(num) < _getScalePosition(upperLimit))
      return outRangeBarColor;
    else if (_getScalePosition(num) == _getScalePosition(midLimitUpper) &&
        _getScalePosition(num) < _getScalePosition(upperLimit))
      return outRangeBarColor;
    else if (_getScalePosition(num) == _getScalePosition(midLimitUpper) &&
        _getScalePosition(num) == _getScalePosition(upperLimit))
      return normalBarColor;
    else
      return normalBarColor;
  }

  /// get color of bigger bar according to range
  Color _getBigBarColor() {
    if (_getScalePosition(num) >= _getScalePosition(lowerLimit) &&
        _getScalePosition(num) < _getScalePosition(midLimitLower))
      return normalBarColor;
    else if (_getScalePosition(num) >= _getScalePosition(midLimitLower) &&
        _getScalePosition(num) <= _getScalePosition(midLimitUpper))
      return inRangeBarColor;
    else if (_getScalePosition(num) > _getScalePosition(midLimitUpper) &&
        _getScalePosition(num) <= _getScalePosition(upperLimit))
      return outRangeBarColor;
    else
      return normalBarColor;
  }

  void _getSmallBars() {
    num = num + 1;

    _show = lowerLimit == num ||
        upperLimit == num ||
        midLimitLower == num ||
        midLimitUpper == num;

    _children = List();

    _children.add(
      Stack(
        alignment: Alignment.topCenter,
        children: <Widget>[
          _show
              ? indicatorWidget
              /*Text(
                  num.toString(),
                  style: TextStyle(
                      fontSize: 14,
                      color: Colors.red,
                      fontWeight: FontWeight.bold),
                )*/
              : Container(),
          Padding(
            padding: const EdgeInsets.only(top: 20.0),
            child: Container(
              margin: EdgeInsets.all(_spacing),
              color: _getBigBarColor(),
              width: 2,
              height: 20,
            ),
          ),
          Padding(
            padding: const EdgeInsets.only(top: 50.0),
            child: Text(
              num.toString(),
              style: TextStyle(
                  fontSize: 14,
                  color: normalBarColor,
                  fontWeight: FontWeight.bold),
            ),
          ),
        ],
      ),
    );

    for (int i = 1; i <= midInterval; i++) {
      _children.add(Padding(
        padding: const EdgeInsets.only(top: 20.0),
        child: Container(
          margin: EdgeInsets.all(_spacing),
          color: _getSmallBarColor(),
          width: 2,
          height: 10,
        ),
      ));
    }
  }

  _getScalePosition(int num) {
    return num + (num - 1) * 3;
  }
}

enum RulerBar { BIG_BAR, SMALL_BAR }


lib/src/ruler_widget.dart
import 'package:flutter/material.dart';

class RulerWidget extends StatefulWidget {
  /// total visible numbers on scale
  int largeScaleBarsInterval;

  /// number of small bars b/w two [largeScaleBarsInterval] on scale
  int smallScaleBarsInterval;

  /// starting number on scale to show marker [indicatorWidget]
  int lowerIndicatorLimit;

  /// ending number on scale  to show marker [indicatorWidget]
  final int upperIndicatorLimit;

  /// mid starting number on scale
  int lowerMidIndicatorLimit;

  /// mid ending number on scale
  int upperMidIndicatorLimit;



  /// color of bars on scale [largeScaleBarsInterval] and [smallScaleBarsInterval]
  final Color barsColor;

  /// color between [lowerMidIndicatorLimit] and [upperMidIndicatorLimit] bars of scale
  final Color inRangeBarColor;

  /// color between [upperMidIndicatorLimit] and [upperIndicatorLimit] bars of scale (Optional)
  final Color outRangeBarColor;

  /// color between [lowerMidIndicatorLimit] and [lowerIndicatorLimit] bars of scale (Optional)
  final Color behindRangeBarColor;

  /// color of scale
  Color scaleBackgroundColor;

  /// Custom Indicator over bars on scale
  Widget indicatorWidget = Container();

  /// height of scale if horizontal or width of scale if vertical
  final double height;

  /// Scale to be horizontal or vertical ,by default horizontal
  final Axis axis;

  final ScrollController? scrollController;

  RulerWidget(
      {Key? key,
      required this.height,
      required this.largeScaleBarsInterval,
      required this.smallScaleBarsInterval,
      required this.scaleBackgroundColor,
      required this.barsColor,
      this.indicatorWidget = const SizedBox(),
      this.lowerIndicatorLimit = 0,
      this.lowerMidIndicatorLimit = 0,
      this.upperMidIndicatorLimit = 0,
      this.upperIndicatorLimit = 0,
      this.inRangeBarColor = Colors.black,
      this.behindRangeBarColor = Colors.black,
      this.outRangeBarColor = Colors.black,
      this.scrollController,
      this.axis = Axis.horizontal})
      : super(key: key);

  @override
  _RulerWidgetState createState() => _RulerWidgetState();
}

class _RulerWidgetState extends State<RulerWidget> {
  late List<Widget> scaleWidgetList;

  @override
  void initState() {
    super.initState();

    scaleWidgetList = [];
    _generateScale();
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      color: widget.scaleBackgroundColor,
      height: widget.axis == Axis.horizontal ? widget.height : null,
      width: widget.axis == Axis.vertical ? widget.height : null,
      child: SingleChildScrollView(
        controller: widget.scrollController,
        scrollDirection: widget.axis,
        child: RotatedBox(
          quarterTurns: widget.axis == Axis.horizontal ? 4 : 3,
          child: Row(
            children: scaleWidgetList,
          ),
        ),
      ),
    );
  }

  void _generateScale() async {
    for (int i = 0; i < widget.largeScaleBarsInterval; i++) {
      print("_generateScale $i");
      scaleWidgetList.add(RulerBarWidget(
        key: ValueKey(i),
        num: i,
        type: RulerBar.BIG_BAR,
        lowerLimit: widget.lowerIndicatorLimit,
        upperLimit: widget.upperIndicatorLimit,
        indicatorWidget: widget.indicatorWidget,
        midLimitLower: widget.lowerMidIndicatorLimit,
        midLimitUpper: widget.upperMidIndicatorLimit,
        midInterval: widget.smallScaleBarsInterval,
        normalBarColor: widget.barsColor,
        inRangeBarColor: widget.inRangeBarColor,
        behindRangeBarColor: widget.behindRangeBarColor,
        outRangeBarColor: widget.outRangeBarColor,
        axis: widget.axis,
      ));
    }

    if (mounted) setState(() {});
  }
}

class RulerBarWidget extends StatelessWidget {
  final RulerBar type;

  int num;

  int midInterval;
  int lowerLimit;
  int upperLimit;
  int midLimitLower;
  int midLimitUpper;
  Color normalBarColor;
  Color inRangeBarColor;
  Color outRangeBarColor;
  Color behindRangeBarColor;
  Widget indicatorWidget;
  Axis axis;

  RulerBarWidget(
      {Key? key,
      required this.num,
      required this.type,
      required this.indicatorWidget,
      required this.lowerLimit,
      required this.midLimitLower,
      required this.midLimitUpper,
      required this.upperLimit,
      required this.midInterval,
      required this.normalBarColor,
      required this.inRangeBarColor,
      required this.behindRangeBarColor,
      required this.outRangeBarColor,
      required this.axis})
      : super(key: key);

  final double _spacing = 5.0;
  List<Widget> _children = List.empty(growable: true);
  bool _show = false;

  @override
  Widget build(BuildContext context) {
    if (_children.isEmpty) _getSmallBars();

    return Container(
      child: Row(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: _children,
      ),
    );
  }

  /// get color of small bar according to range
  Color _getSmallBarColor() {
    if (_getScalePosition(num) >= _getScalePosition(lowerLimit) &&
        _getScalePosition(num) < _getScalePosition(midLimitLower))
      return behindRangeBarColor;
    else if (_getScalePosition(num) >= _getScalePosition(midLimitLower) &&
        _getScalePosition(num) < _getScalePosition(midLimitUpper))
      return inRangeBarColor;
    else if (_getScalePosition(num) > _getScalePosition(midLimitUpper) &&
        _getScalePosition(num) < _getScalePosition(upperLimit))
      return outRangeBarColor;
    else if (_getScalePosition(num) == _getScalePosition(midLimitUpper) &&
        _getScalePosition(num) < _getScalePosition(upperLimit))
      return outRangeBarColor;
    else if (_getScalePosition(num) == _getScalePosition(midLimitUpper) &&
        _getScalePosition(num) == _getScalePosition(upperLimit))
      return normalBarColor;
    else
      return normalBarColor;
  }

  /// get color of bigger bar according to range
  Color _getBigBarColor() {
    if (_getScalePosition(num) >= _getScalePosition(lowerLimit) &&
        _getScalePosition(num) < _getScalePosition(midLimitLower))
      return normalBarColor;
    else if (_getScalePosition(num) >= _getScalePosition(midLimitLower) &&
        _getScalePosition(num) <= _getScalePosition(midLimitUpper))
      return inRangeBarColor;
    else if (_getScalePosition(num) > _getScalePosition(midLimitUpper) &&
        _getScalePosition(num) <= _getScalePosition(upperLimit))
      return outRangeBarColor;
    else
      return normalBarColor;
  }

  void _getSmallBars() {
    num = num + 1;

    _show = lowerLimit == num ||
        upperLimit == num ||
        midLimitLower == num ||
        midLimitUpper == num;

    _children = List.empty(growable: true);

    _children.add(
      Stack(
        alignment: Alignment.topCenter,
        children: <Widget>[
          _show ? indicatorWidget : Container(),
          Padding(
            padding: const EdgeInsets.only(top: 20.0),
            child: Container(
              margin: EdgeInsets.all(_spacing),
              color: _getBigBarColor(),
              width: 2,
              height: 20,
            ),
          ),
          Padding(
            padding: const EdgeInsets.only(top: 50.0),
            child: Text(
              num.toString(),
              style: TextStyle(
                  fontSize: 14,
                  color: normalBarColor,
                  fontWeight: FontWeight.bold),
            ),
          ),
        ],
      ),
    );

    for (int i = 1; i <= midInterval; i++) {
      _children.add(Padding(
        padding: const EdgeInsets.only(top: 20.0),
        child: Container(
          margin: EdgeInsets.all(_spacing),
          color: _getSmallBarColor(),
          width: 2,
          height: 10,
        ),
      ));
    }
  }

  _getScalePosition(int num) {
    return num + (num - 1) * 3;
  }
}

enum RulerBar { BIG_BAR, SMALL_BAR }


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:rulers/rulers.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
          primarySwatch: Colors.blue,
          textTheme: TextTheme(
              headline3: TextStyle(
                  color: Colors.black,
                  fontWeight: FontWeight.bold,
                  fontSize: 20))),
      home: Demo(),
    );
  }
}

class Demo extends StatefulWidget {
  @override
  _DemoState createState() => _DemoState();
}

/*
As for Example only here
              * Limit is 24 as time in hours
              * interval 3 as 15,30,45 min
              *
              *
              * */

class _DemoState extends State<Demo> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Rulers Example'),
      ),
      body: Padding(
        padding: const EdgeInsets.all(8.0),
        child: SingleChildScrollView(
          child: Padding(
            padding: const EdgeInsets.only(top: 20.0),
            child: Column(
              mainAxisSize: MainAxisSize.min,
              children: <Widget>[
                Text(
                  'Normal Scale',
                  style: Theme.of(context).textTheme.headline3,
                ),
                Container(
                  margin: const EdgeInsets.only(top: 8.0),
                  child: RulerWidget(
                    height: 100,
                    scaleBackgroundColor: Colors.grey.shade200,
                    largeScaleBarsInterval: 24,
                    smallScaleBarsInterval: 3,
                    barsColor: Colors.grey,
                    indicatorWidget: Column(
                      children: <Widget>[
                        Icon(
                          Icons.arrow_drop_down,
                          color: Colors.red,
                        ),
                      ],
                    ),
                  ),
                ),
                Padding(
                  padding: const EdgeInsets.only(top: 20.0),
                  child: Text(
                    'In and Out Range Scale',
                    style: Theme.of(context).textTheme.headline3,
                  ),
                ),
                Container(
                  margin: const EdgeInsets.only(top: 8.0),
                  child: RulerWidget(
                    height: 100,
                    scaleBackgroundColor: Colors.grey.shade100,
                    indicatorWidget: Column(
                      children: const <Widget>[
                        Icon(
                          Icons.arrow_drop_down,
                          color: Colors.red,
                        ),
                      ],
                    ),
                    largeScaleBarsInterval: 24,
                    smallScaleBarsInterval: 3,
                    lowerIndicatorLimit: 2,
                    lowerMidIndicatorLimit: 2,
                    upperMidIndicatorLimit: 6,
                    upperIndicatorLimit: 6,
                    barsColor: Colors.grey,
                    inRangeBarColor: Colors.green,
                    behindRangeBarColor: Colors.orangeAccent,
                    outRangeBarColor: Colors.red,
                  ),
                ),
                Padding(
                  padding: const EdgeInsets.only(top: 20.0),
                  child: Text(
                    'In,Out,Mid Range Scale',
                    style: Theme.of(context).textTheme.headline3,
                  ),
                ),
                Container(
                  margin: const EdgeInsets.only(top: 8.0),
                  child: RulerWidget(
                    height: 100,
                    scaleBackgroundColor: Colors.yellow,
                    indicatorWidget: Column(
                      children: const <Widget>[
                        Icon(
                          Icons.arrow_drop_down,
                          color: Colors.red,
                        ),
                      ],
                    ),
                    largeScaleBarsInterval: 24,
                    smallScaleBarsInterval: 3,
                    lowerIndicatorLimit: 2,
                    lowerMidIndicatorLimit: 4,
                    upperMidIndicatorLimit: 7,
                    upperIndicatorLimit: 8,
                    barsColor: Colors.grey,
                    inRangeBarColor: Colors.green,
                    behindRangeBarColor: Colors.orangeAccent,
                    outRangeBarColor: Colors.red,
                  ),
                ),
                Padding(
                  padding: const EdgeInsets.only(top: 20.0),
                  child: Text(
                    'Custom Marker Scale',
                    style: Theme.of(context).textTheme.headline3,
                  ),
                ),
                Container(
                  margin: const EdgeInsets.only(top: 8.0),
                  child: RulerWidget(
                    height: 100,
                    scaleBackgroundColor: Colors.blue.shade100,
                    indicatorWidget: Image.asset(
                      'assets/marker.png',
                      color: Colors.red,
                      height: 25,
                      width: 25,
                    ),
                    largeScaleBarsInterval: 24,
                    smallScaleBarsInterval: 3,
                    lowerIndicatorLimit: 2,
                    lowerMidIndicatorLimit: 2,
                    upperMidIndicatorLimit: 7,
                    upperIndicatorLimit: 7,
                    barsColor: Colors.grey,
                    inRangeBarColor: Colors.green,
                    behindRangeBarColor: Colors.orangeAccent,
                    outRangeBarColor: Colors.red,
                  ),
                ),
                Padding(
                  padding: const EdgeInsets.only(top: 20.0),
                  child: Text(
                    'Horizontal Scale',
                    style: Theme.of(context).textTheme.headline3,
                  ),
                ),
                Container(
                  margin: const EdgeInsets.only(top: 8.0),
                  alignment: Alignment.centerLeft,
                  child: RulerWidget(
                    scaleBackgroundColor: Colors.grey.shade100,
                    height: 100,
                    indicatorWidget: Column(
                      children: <Widget>[
                        Icon(
                          Icons.arrow_drop_down,
                          color: Colors.red,
                        ),
                      ],
                    ),
                    largeScaleBarsInterval: 20,
                    smallScaleBarsInterval: 4,
                    lowerIndicatorLimit: 0,
                    lowerMidIndicatorLimit: 0,
                    upperMidIndicatorLimit: 0,
                    upperIndicatorLimit: 0,
                    barsColor: Colors.grey,
                    inRangeBarColor: Colors.green,
                    behindRangeBarColor: Colors.orangeAccent,
                    outRangeBarColor: Colors.red,
                    axis: Axis.vertical,
                  ),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}
